File: src/main/java/com/herobrine/future/item/ItemTrident.java
package com.herobrine.future.item;

import com.google.common.collect.Multimap;
import com.herobrine.future.FutureMC;
import com.herobrine.future.client.Modeled;
import com.herobrine.future.enchantment.EnchantHelper;
import com.herobrine.future.entity.trident.EntityTrident;
import com.herobrine.future.init.FutureConfig;
import com.herobrine.future.sound.Sounds;
import net.minecraft.block.state.IBlockState;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.MoverType;
import net.minecraft.entity.SharedMonsterAttributes;
import net.minecraft.entity.ai.attributes.AttributeModifier;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.projectile.EntityArrow;
import net.minecraft.inventory.EntityEquipmentSlot;
import net.minecraft.item.EnumAction;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ActionResult;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumHand;
import net.minecraft.util.SoundCategory;
import net.minecraft.util.SoundEvent;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.MathHelper;
import net.minecraft.world.World;

public class ItemTrident extends Item implements Modeled {
    public ItemTrident() {
        setRegistryName("Trident");
        setUnlocalizedName(FutureMC.ID + ".Trident");
        setCreativeTab(FutureConfig.general.useVanillaTabs ? CreativeTabs.COMBAT : FutureMC.CREATIVE_TAB);
        setMaxDamage(250);
        setMaxStackSize(1);
        addModel();
    }

    @Override
    public boolean canDestroyBlockInCreative(World world, BlockPos pos, ItemStack stack, EntityPlayer player) {
        return false;
    }

    @Override
    public EnumAction getItemUseAction(ItemStack stack) {
        return EnumAction.BOW;
    }

    @Override
    public int getMaxItemUseDuration(ItemStack stack) {
        return 72000;
    }

    @Override
    public boolean hasEffect(ItemStack stack) {
        return false;
    }

    @Override
    public void onPlayerStoppedUsing(ItemStack stack, World worldIn, EntityLivingBase entityLiving, int timeLeft) {
        if (entityLiving instanceof EntityPlayer) {
            EntityPlayer entityPlayer = (EntityPlayer)entityLiving;
            int i = getUseDuration(stack, timeLeft);
            if (i >= 10) {
                int j = EnchantHelper.getRiptide(stack);
                if(j <= 0 || entityPlayer.isWet()) {
                    if (!worldIn.isRemote) {
                        stack.damageItem(1, entityPlayer);
                        if(j == 0) {
                            EntityTrident trident = new EntityTrident(worldIn, entityPlayer, stack);
                            trident.shoot(entityLiving, entityPlayer.rotationPitch, entityPlayer.rotationYaw, 0.0F, 2.5F + (float)j * 0.5F, 1.0F);

                            if (entityPlayer.capabilities.isCreativeMode) {
                                trident.pickupStatus = EntityArrow.PickupStatus.CREATIVE_ONLY;
                            }

                            worldIn.spawnEntity(trident);
                            if (!entityPlayer.capabilities.isCreativeMode) {
                                entityPlayer.inventory.deleteStack(stack);
                            }
                        }
                    }

                    SoundEvent soundevent = Sounds.TRIDENT_THROW;
                    if (j > 0) {
                        float f = entityPlayer.rotationYaw;
                        float f1 = entityPlayer.rotationPitch;
                        float f2 = -MathHelper.sin(f * ((float)Math.PI / 180F)) * MathHelper.cos(f1 * ((float)Math.PI / 180F));
                        float f3 = -MathHelper.sin(f1 * ((float)Math.PI / 180F));
                        float f4 = MathHelper.cos(f * ((float)Math.PI / 180F)) * MathHelper.cos(f1 * ((float)Math.PI / 180F));
                        float f5 = MathHelper.sqrt(f2 * f2 + f3 * f3 + f4 * f4);
                        float f6 = 3.0F * ((1.0F + (float)j) / 4.0F);
                        f2 = f2 * (f6 / f5);
                        f3 = f3 * (f6 / f5);
                        f4 = f4 * (f6 / f5);
                        entityPlayer.addVelocity(f2, f3, f4);
                        if (j >= 3) {
                            soundevent = Sounds.TRIDENT_RIPTIDE_2;
                        } else if (j == 2) {
                            soundevent = Sounds.TRIDENT_RIPTIDE_1;
                        } else {
                            soundevent = Sounds.TRIDENT_RIPTIDE_0;
                        }

                        if (entityPlayer.onGround) {
                            entityPlayer.move(MoverType.SELF, 0.0D, (double)1.1999999F, 0.0D);
                        }
                    }

                    worldIn.playSound(null, entityPlayer.posX, entityPlayer.posY, entityPlayer.posZ, soundevent, SoundCategory.PLAYERS, 1.0F, 1.0F);
                }
            }
        }
    }

    @Override
    public ActionResult<ItemStack> onItemRightClick(World worldIn, EntityPlayer playerIn, EnumHand hand) {
        ItemStack itemstack = playerIn.getHeldItem(hand);
        if (itemstack.getItemDamage() >= itemstack.getMaxDamage()) {
            return new ActionResult<>(EnumActionResult.FAIL, itemstack);
        } else if (EnchantHelper.getRiptide(itemstack) > 0 && !playerIn.isWet()) {
            return new ActionResult<>(EnumActionResult.FAIL, itemstack);
        } else {
            playerIn.setActiveHand(hand);
            return new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
        }
    }

    @Override
    public boolean hitEntity(ItemStack stack, EntityLivingBase target, EntityLivingBase attacker) {
        stack.damageItem(1, attacker);
        return true;
    }

    private int getUseDuration(ItemStack stack, int timeLeft) {
        return stack.getMaxItemUseDuration() - timeLeft;
    }

    @Override
    public boolean onBlockDestroyed(ItemStack stack, World worldIn, IBlockState state, BlockPos pos, EntityLivingBase entityLiving) {
        if (state.getBlockHardness(worldIn, pos) != 0.0D) {
            stack.damageItem(2, entityLiving);
        }

        return true;
    }

    @Override
    public Multimap<String, AttributeModifier> getItemAttributeModifiers(EntityEquipmentSlot equipmentSlot) {
        Multimap<String, AttributeModifier> multimap = super.getItemAttributeModifiers(equipmentSlot);

        if (equipmentSlot == EntityEquipmentSlot.MAINHAND) {
            multimap.put(SharedMonsterAttributes.ATTACK_DAMAGE.getName(), new AttributeModifier(ATTACK_DAMAGE_MODIFIER, "Weapon modifier", 8.0D, 0));
            multimap.put(SharedMonsterAttributes.ATTACK_SPEED.getName(), new AttributeModifier(ATTACK_SPEED_MODIFIER, "Weapon modifier", (double)-2.9F, 0));
        }

        return multimap;
    }

    public int getItemEnchantability() {
        return 1;
    }

    @Override
    public int getItemStackLimit() {
        return 1;
    }
}

File: src/main/java/thedarkcolour/futuremc/item/ItemTrident.kt
package thedarkcolour.futuremc.item

import com.google.common.collect.Multimap
import net.minecraft.block.state.IBlockState
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.MoverType
import net.minecraft.entity.SharedMonsterAttributes
import net.minecraft.entity.ai.attributes.AttributeModifier
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.projectile.EntityArrow
import net.minecraft.inventory.EntityEquipmentSlot
import net.minecraft.item.EnumAction
import net.minecraft.item.ItemStack
import net.minecraft.util.ActionResult
import net.minecraft.util.EnumActionResult
import net.minecraft.util.EnumHand
import net.minecraft.util.SoundCategory
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import thedarkcolour.core.item.ItemModeled
import thedarkcolour.futuremc.FutureMC
import thedarkcolour.futuremc.config.FConfig
import thedarkcolour.futuremc.enchantment.EnchantHelper
import thedarkcolour.futuremc.entity.trident.EntityTrident
import thedarkcolour.futuremc.registry.FSounds
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

class ItemTrident : ItemModeled("Trident") {
    override fun canDestroyBlockInCreative(
        world: World,
        pos: BlockPos,
        stack: ItemStack,
        player: EntityPlayer
    ): Boolean {
        return false
    }

    override fun getItemUseAction(stack: ItemStack): EnumAction = EnumAction.BOW

    override fun getMaxItemUseDuration(stack: ItemStack): Int = 72000

    override fun hasEffect(stack: ItemStack): Boolean = false

    override fun isEnchantable(stack: ItemStack): Boolean = true

    override fun onPlayerStoppedUsing(stack: ItemStack, worldIn: World, entityLiving: EntityLivingBase, timeLeft: Int) {
        if (entityLiving is EntityPlayer) {
            val i = getUseDuration(stack, timeLeft)
            if (i >= 10) {
                val j = EnchantHelper.getRiptide(stack)
                if (j <= 0 || entityLiving.isWet) {
                    if (!worldIn.isRemote) {
                        stack.damageItem(1, entityLiving)
                        if (j == 0) {
                            val trident = EntityTrident(worldIn, entityLiving, stack)
                            trident.shoot(
                                entityLiving,
                                entityLiving.rotationPitch,
                                entityLiving.rotationYaw,
                                0.0f,
                                2.5f + j.toFloat() * 0.5f,
                                1.0f
                            )
                            if (entityLiving.capabilities.isCreativeMode) {
                                trident.pickupStatus = EntityArrow.PickupStatus.CREATIVE_ONLY
                            }
                            worldIn.spawnEntity(trident)
                            if (!entityLiving.capabilities.isCreativeMode) {
                                entityLiving.inventory.deleteStack(stack)
                            }
                        }
                    }
                    var sound = FSounds.TRIDENT_THROW
                    if (j > 0) {
                        val f = entityLiving.rotationYaw
                        val f1 = entityLiving.rotationPitch
                        var f2 = -sin(f * (PI.toFloat() / 180f)) * cos(f1 * (PI.toFloat() / 180f))
                        var f3 = -sin(f1 * (PI.toFloat() / 180f))
                        var f4 = cos(f * (PI.toFloat() / 180f)) * cos(f1 * (PI.toFloat() / 180f))
                        val f5 = sqrt(f2 * f2 + f3 * f3 + f4 * f4)
                        val f6 = 3.0f * ((1.0f + j.toFloat()) / 4.0f)
                        f2 *= (f6 / f5)
                        f3 *= (f6 / f5)
                        f4 *= (f6 / f5)
                        entityLiving.addVelocity(f2.toDouble(), f3.toDouble(), f4.toDouble())
                        sound = when {
                            j >= 3 -> {
                                FSounds.TRIDENT_RIPTIDE_III
                            }
                            j == 2 -> {
                                FSounds.TRIDENT_RIPTIDE_II
                            }
                            else -> {
                                FSounds.TRIDENT_RIPTIDE_I
                            }
                        }
                        if (entityLiving.onGround) {
                            entityLiving.move(MoverType.SELF, 0.0, 1.1999999, 0.0)
                        }
                    }
                    worldIn.playSound(
                        null,
                        entityLiving.posX,
                        entityLiving.posY,
                        entityLiving.posZ,
                        sound,
                        SoundCategory.PLAYERS,
                        1.0f,
                        1.0f
                    )
                }
            }
        }
    }

    override fun onItemRightClick(worldIn: World, playerIn: EntityPlayer, hand: EnumHand): ActionResult<ItemStack> {
        val stack = playerIn.getHeldItem(hand)
        return if (stack.itemDamage >= stack.maxDamage) {
            ActionResult(EnumActionResult.FAIL, stack)
        } else if (EnchantHelper.getRiptide(stack) > 0 && !playerIn.isWet) {
            ActionResult(EnumActionResult.FAIL, stack)
        } else {
            playerIn.activeHand = hand
            ActionResult(EnumActionResult.SUCCESS, stack)
        }
    }

    override fun hitEntity(stack: ItemStack, target: EntityLivingBase, attacker: EntityLivingBase): Boolean {
        stack.damageItem(1, attacker)
        return true
    }

    private fun getUseDuration(stack: ItemStack, timeLeft: Int): Int {
        return stack.maxItemUseDuration - timeLeft
    }

    override fun onBlockDestroyed(
        stack: ItemStack,
        worldIn: World,
        state: IBlockState,
        pos: BlockPos,
        entityLiving: EntityLivingBase
    ): Boolean {
        if (state.getBlockHardness(worldIn, pos).toDouble() != 0.0) {
            stack.damageItem(2, entityLiving)
        }
        return true
    }

    override fun getItemAttributeModifiers(equipmentSlot: EntityEquipmentSlot): Multimap<String, AttributeModifier> {
        val multimap = super.getItemAttributeModifiers(equipmentSlot)
        if (equipmentSlot == EntityEquipmentSlot.MAINHAND) {
            multimap.put(
                SharedMonsterAttributes.ATTACK_DAMAGE.name,
                AttributeModifier(ATTACK_DAMAGE_MODIFIER, "Weapon modifier", 8.0, 0)
            )
            multimap.put(
                SharedMonsterAttributes.ATTACK_SPEED.name,
                AttributeModifier(ATTACK_SPEED_MODIFIER, "Weapon modifier", (-2.9f).toDouble(), 0)
            )
        }
        return multimap
    }

    override fun getItemEnchantability(): Int = 1

    override fun getItemStackLimit(): Int = 1

    //val tileEntityItemStackRendererr: TileEntityItemStackRenderer
    //    get() = SpecialItemRenderer.instance

    init {
        creativeTab = if (FConfig.useVanillaCreativeTabs) CreativeTabs.COMBAT else FutureMC.TAB
        maxDamage = 250
        setMaxStackSize(1)
    }
}